[14기 팝콘]step1 카페메뉴 만들기

yeomgahui opened the pull request:
STEP 1 구현을 마쳤습니다.

개인적으로 구현하면서 아쉬웠던 부분을 생각해보면,
1. menu 추가시 수정 삭제를 위해 addEventListener를 등록 하는데, 메뉴 하나를 추가할때마다 모든 추가, 삭제 버튼의 eventListener를 다시 등록하도록 구현했습니다. => 개인적으로는 하나 추가할때마다 해당 수정, 삭제 버튼의  eventLIstener만을 추가하도록 하고 싶었는데 좋은 방법이 떠오르지 않았습니다. 
2. JavaScript 파일을 기능 마다 분리하여서 작성하고 싶었는데, 어떻게 나눌지 모르곘어서 index.js 파일 내에 모든 로직을 다 작성했습니다. 이부분은 차차 다시 생각해보고 수정해보고 싶습니다.

yeomgahui committed:
Update README.md

yeomgahui committed:
step1 구현완료

dhrod0325 commented on src/js/index.js:
```diff
@@ -0,0 +1,73 @@
+const totalCount = document.querySelector('.menu-count');
+
+inputField.addEventListener("keypress", handleEnter);
+submitButton.addEventListener("click", handleSubmit);
```
keypress로 enter를 확인하고 확인 버튼을 눌렀을 때를 체크하는 방법도 있지만 button을 submit으로 만들고 html자체적으로 enter이벤트를 핸들링 하는 방법도 있어요 😊

예를 들면 다음과 같이요!

```js
<form id="form">
<input type="text"/>
<button type="submit">저장</button>
</form>

<script>
const $form = document.getElementById('form');

$form.addEventListener('submit', e=>{
  e.preventDefault();
 
  //submit 했을 때 동작
});
</script>
```

위와 같이 html을 작성하면 enter이벤트를 따로 주시지 않아도 되고 js가 동작하지 않을 때에도 enter이벤트가 동작해요~!

참고 하시면 좋을 것 같아요!

yeomgahui committed:
[14기 팝콘] Step1 Comment 추가

STEP 1 구현을 마쳤습니다.

개인적으로 구현하면서 아쉬웠던 부분을 생각해보면,

menu 추가시 수정 삭제를 위해 addEventListener를 등록 하는데, 메뉴 하나를 추가할때마다 모든 추가, 삭제 버튼의 eventListener를 다시 등록하도록 구현했습니다. => 개인적으로는 하나 추가할때마다 해당 수정, 삭제 버튼의 eventLIstener만을 추가하도록 하고 싶었는데 좋은 방법이 떠오르지 않았습니다.
JavaScript 파일을 기능 마다 분리하여서 작성하고 싶었는데, 어떻게 나눌지 모르곘어서 index.js 파일 내에 모든 로직을 다 작성했습니다. 이부분은 차차 다시 생각해보고 수정해보고 싶습니다.

dhrod0325 commented on src/js/index.js:
```diff
@@ -0,0 +1,73 @@
+const inputField = document.getElementById('espresso-menu-name');
+const submitButton = document.getElementById('espresso-menu-submit-button');
+const menuList = document.getElementById('espresso-menu-list');
+const totalCount = document.querySelector('.menu-count');
```
사용되는 dom을 상단에 놓는거 참 좋은 것 같아요! 👍

dahye1013 replied:
👍 모듈 분리하고 싶다고하셔서 덧붙이면 `selector` 요 친구들을 분리해봐도 좋을 것 같네요.
`selector`용 유틸리티를 만들어도 좋을 것 같아용

dhrod0325 commented on src/js/index.js:
```diff
@@ -0,0 +1,73 @@
+}
+
+function addMenu(item){
+    menuList.innerHTML += 
```
`innerHTML`로 html을 붙여 나가는 식으로 구현 하셨네요.! 

inenerHTML은 붙여넣을 dom을 파싱하고 붙여넣을 string도 파싱해서 element로 변환하는 작업을 해야하기 때문에 성능상으로 비효율적인 측면이 있어요!

`insertAdjacentHTML` 함수로 변경 해보는 건 어떨까요~?

아래 링크도 한번 읽어보시면 좋을 것 같아요!

[innerHTML과 insertAdjacentHTML](https://usage.tistory.com/133)

dhrod0325 commented on src/js/index.js:
```diff
@@ -0,0 +1,73 @@
+    editButtons.forEach((editButton, index) => {
+        editButton.addEventListener('click', handleEdit);
+        removeButtons[index].addEventListener('click', handleDelete);
+    });
```
PR에 남겨주신 내용이 이 부분인 것 같아요.

지금은 li 태그를 추가 할 때 마다 전체 수정 삭제 버튼을 모두 가지고 와서 이벤트를 다시 바인딩 해주고 있는데 만약에 메뉴가 아니고 몇백 또는 몇천개의 element가 추가되고 있는 상황이라면 하나 추가 할 때마다 전체 이벤트를 다시 바인딩 해주어야 하는데 처리하는 시간이 엄청 차이날 것 같아요

 innerHTML로 element 전체를 다시 만들고 있는 상황에 연결되는 부분이라서 다시 바인딩 하는 방법 말고는 없어 보이네요. li 태그를 append하면서 하나씩 하나씩 바인딩 해주는 식으로 변경 해보면 어떨까요~?

dhrod0325 commented on src/js/index.js:
```diff
@@ -0,0 +1,73 @@
+}
+
+function handleEdit(event){
+    const menuName = event.target.previousElementSibling;
```
`previousElementSibling` 같이 html 태그에 의존적으로 js코드를 작성하면 html에 element 위치를 변경하는 것 만으로도 코드가 동작을 하지 않게 되니 다른 방법으로 변경해보면 어때요~? 아래 코드에 있는 `childNodes` 나 `parentNode` 같은 부분도 약간 변경 해 볼 수 있을 것 같아요

dhrod0325 reviewed (commented):
안녕하세요 팝콘님~! 

변수명도 코드 의도를 알아보기 쉽고 함수도 분리가 정말 잘 되어있어서 읽기 정말 좋았어요~!

PR에 작성하신대로 전체 이벤트를 다시 바인딩 하는 부분만 변경되면 정말 좋을 것 같아요!

몇 가지 제 생각을 남겨 보았어요~! 정말 고생하셨습니닷! 

남은 미션도 같이 열심히 해봐용! 😊

dhrod0325 commented on src/js/index.js:
```diff
@@ -0,0 +1,73 @@
+function updateTotalCount(){
+    const count = menuList.childNodes.length;
+    totalCount.innerText = `총 ${count}개`;
+} 
```
여기 `eol` 제거 하는 부분은 뛰어쓰기보단 개행하는게 더 좋을 것 같아요~!

아래 링크도 한번 읽어보시면 좋을 것 같습니다~!

https://coderifleman.tumblr.com/post/115464362564/%ED%8C%8C%EC%9D%BC-%EB%81%9D%EC%97%90-%EA%B0%9C%ED%96%89%EC%9D%84-%EC%B6%94%EA%B0%80%ED%95%B4%EC%95%BC-%ED%95%98%EB%8A%94-%EC%9D%B4%EC%9C%A0

dahye1013 commented on src/js/index.js:
```diff
@@ -0,0 +1,73 @@
+  </li>`;
+    setItemEvent();
+}
+function setItemEvent(){
+    const editButtons = document.querySelectorAll('.menu-edit-button');
+    const removeButtons = document.querySelectorAll('.menu-remove-button');
```
event delegation을 사용하는 건 어떨까요 ㅎㅎㅎ?
section처럼 좀 더 큰 단위에 이벤트를 붙여놓으면 따로 element가 추가 될 때마다 이벤트를 붙이지 않아도 될 것 같네용
https://ko.javascript.info/event-delegation

yeomgahui committed:
test

yeomgahui committed:
test

yeomgahui committed:
test

yeomgahui committed:
필요없는 파일 제거

yeomgahui committed:
[14기 팝콘] 코드리뷰 후 수정

yeomgahui committed:
[Step 1] 필요없는 주석 제거
